src/System.Private.ServiceModel/src/System/ServiceModel/Security/SecurityProtocolFactory.cs

```diff
@@ -686,6 +693,31 @@ public virtual void OnClose(TimeSpan timeout)
             }
         }
 
+        public virtual Task OnCloseAsync(TimeSpan timeout)
+        {
+            return OnCloseAsyncInternal(timeout);
+        }
+
+        private async Task OnCloseAsyncInternal(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            if (!_actAsInitiator)
+            {
+                foreach (SupportingTokenAuthenticatorSpecification spec in _channelSupportingTokenAuthenticatorSpecification)
+                {
+                    await SecurityUtils.CloseTokenAuthenticatorIfRequiredAsync(spec.TokenAuthenticator, timeoutHelper.RemainingTime());
+                }
+                foreach (string action in _scopedSupportingTokenAuthenticatorSpecification.Keys)
```

Comment:
Yeah, I was constantly tempted to refactor, but in the end did a verbatim copy and then Taskified it.  I sort of prefer keeping them looking as similar as possible so we'd recognize when one of them changed.

---
In reply to: [78407020](https://github.com/dotnet/wcf/pull/1515#discussion_r78407020) [](ancestors = 78407020,78406802)